plugins/woocommerce/client/blocks/assets/js/extensions/payment-methods/paypal/buttons.js

```diff
@@ -0,0 +1,188 @@
+		disableFunding: disableFunding || '',
+		enableFunding: enableFunding || '',
+		currency: currency || '',
+		intent: intent || '',
+		merchantId: merchantId || '',
+		'data-partner-attribution-id': partnerAttributionId || '',
+		'data-page-type': pageType || '',
+	};
+
+	const createOrder = async ( data ) => {
+		let responseData;
+		try {
+			// If we're inside the product page, we need to empty the cart,
+			// and add the current product to the cart.
+			if ( isProductPage ) {
+				// Empty the cart.
+				await apiFetch( {
+					method: 'DELETE',
+					path: '/wc/store/v1/cart/items',
+				} );
+
+				// Get product ID from the value of the "add-to-cart" button.
+				let productId = document.querySelector(
+					'[name="add-to-cart"]'
+				)?.value;
+				const variationId = document.querySelector(
+					'[name="variation_id"]'
+				)?.value;
+
+				if ( variationId ) {
+					productId = variationId;
+				}
+
+				if ( ! productId ) {
+					return null;
+				}
+
+				// Get quantity from the value of the "quantity" input field.
+				const quantity =
+					document.querySelector( '[name="quantity"]' )?.value;
+				if ( ! quantity ) {
+					return null;
+				}
+
+				// Add the product to the cart.
+				await apiFetch( {
+					method: 'POST',
+					path: '/wc/store/v1/cart/items',
+					data: {
+						id: productId,
+						quantity,
+					},
+				} );
+			}
+
+			// Create a draft order in WooCommerce.
+			responseData = await apiFetch( {
+				method: 'GET',
+				path: '/wc/store/v1/checkout',
+				headers: {
+					Nonce: payPalData.wc_store_api_nonce,
+				},
+			} );
+
+			if ( ! responseData.order_id ) {
+				// eslint-disable-next-line no-console
+				console.error(
+					'Failed to create WooCommerce order',
+					responseData
+				);
+				return null;
+			}
+
+			// Create a PayPal order.
+			const paypalResponseData = await apiFetch( {
+				method: 'POST',
+				path: '/wc/v3/paypal-buttons/create-order',
+				headers: {
+					Nonce: payPalData.create_order_nonce,
+				},
+				data: {
+					order_id: responseData.order_id,
+					payment_source: data.paymentSource || '',
+				},
+			} );
+
+			setOrderId( paypalResponseData.order_id );
+			setOrderReceivedURL( paypalResponseData.return_url );
+
+			return paypalResponseData.paypal_order_id;
+		} catch ( error ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Failed to create order', error );
+			return null;
+		}
+	};
+
+	const onApprove = ( data ) => {
+		if ( data.paymentID && orderReceivedUrl ) {
+			window.location.href = orderReceivedUrl;
+		}
+	};
+
+	const onCancel = async () => {
+		try {
+			await apiFetch( {
+				method: 'POST',
+				path: '/wc/v3/paypal-buttons/cancel-payment',
+				headers: {
+					Nonce: payPalData.cancel_payment_nonce,
+				},
+				data: {
+					order_id: orderId,
+				},
+			} );
+
+			setOrderReceivedURL( '' );
+		} catch ( error ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Failed to cancel PayPal payment', error );
```

Comment:
Does this not need to be shown to the user?

Reply:
The `onError` callback handles displaying the error to the user.

Reply:
Sorry for not being clear in my previous comment. 

All payment-related errors are handled and displayed in the `onError` callback. These errors are meaningful to the user.

This `onCancel` handler is called when the user closes the PayPal payment modal or clicks on the `return to store` link. In this case, we send a request to change the WooCommerce order status. It's okay if we do not show this error message because it is not useful to the user.